File: Kernel Simulator/Misc/Screensaver/Displays/Noise.vb

'    Kernel Simulator  Copyright (C) 2018-2022  Aptivi
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Namespace Misc.Screensaver.Displays
    Public Module NoiseSettings

        Private _noiseNewScreenDelay As Integer = 5000
        Private _noiseDensity As Integer = 40

        ''' <summary>
        ''' [Noise] How many milliseconds to wait before making the new screen?
        ''' </summary>
        Public Property NoiseNewScreenDelay As Integer
            Get
                Return _noiseNewScreenDelay
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 5000
                _noiseNewScreenDelay = value
            End Set
        End Property
        ''' <summary>
        ''' [Noise] The noise density in percent
        ''' </summary>
        Public Property NoiseDensity As Integer
            Get
                Return _noiseDensity
            End Get
            Set(value As Integer)
                If value < 0 Then value = 40
                If value > 100 Then value = 40
                _noiseDensity = value
            End Set
        End Property

    End Module

    Public Class NoiseDisplay
        Inherits BaseScreensaver
        Implements IScreensaver

        Private RandomDriver As Random
        Private CurrentWindowWidth As Integer
        Private CurrentWindowHeight As Integer
        Private ResizeSyncing As Boolean

        Public Overrides Property ScreensaverName As String = "Noise" Implements IScreensaver.ScreensaverName

        Public Overrides Property ScreensaverSettings As Dictionary(Of String, Object) Implements IScreensaver.ScreensaverSettings

        Public Overrides Sub ScreensaverPreparation() Implements IScreensaver.ScreensaverPreparation
            'Variable preparations
            RandomDriver = New Random
            CurrentWindowWidth = ConsoleWrapper.WindowWidth
            CurrentWindowHeight = ConsoleWrapper.WindowHeight
        End Sub

        Public Overrides Sub ScreensaverLogic() Implements IScreensaver.ScreensaverLogic
            Dim NoiseDense As Double = If(NoiseDensity > 100, 100, NoiseDensity) / 100

            Console.BackgroundColor = ConsoleColor.DarkGray
            ConsoleWrapper.CursorVisible = False
            ConsoleWrapper.Clear()
            Console.BackgroundColor = ConsoleColor.Black

            'Select random positions to generate noise
            Dim AmountOfBlocks As Integer = ConsoleWrapper.WindowWidth * ConsoleWrapper.WindowHeight
            Dim BlocksToCover As Integer = AmountOfBlocks * NoiseDense
            Dim CoveredBlocks As New ArrayList
            Do Until CoveredBlocks.Count = BlocksToCover Or ResizeSyncing
                If CurrentWindowHeight <> ConsoleWrapper.WindowHeight Or CurrentWindowWidth <> ConsoleWrapper.WindowWidth Then ResizeSyncing = True
                If Not ResizeSyncing Then
                    Dim CoverX As Integer = RandomDriver.Next(ConsoleWrapper.WindowWidth)
                    Dim CoverY As Integer = RandomDriver.Next(ConsoleWrapper.WindowHeight)
                    ConsoleWrapper.SetCursorPosition(CoverX, CoverY)
                    WritePlain(" ", False)
                    If Not CoveredBlocks.Contains(CStr(CoverX) + ", " + CStr(CoverY)) Then CoveredBlocks.Add(CStr(CoverX) + ", " + CStr(CoverY))
                Else
                    'We're resizing.
                    Exit Do
                End If
            Loop

            'Reset resize sync
            ResizeSyncing = False
            CurrentWindowWidth = ConsoleWrapper.WindowWidth
            CurrentWindowHeight = ConsoleWrapper.WindowHeight
            SleepNoBlock(NoiseNewScreenDelay, ScreensaverDisplayerThread)
        End Sub

    End Class
End Namespace

File: Kernel Simulator/Misc/Screensaver/Displays/Noise.cs
using System;
using System.Collections;
using System.Collections.Generic;
using KS.Misc.Threading;
using KS.Misc.Writers.ConsoleWriters;
using Terminaux.Base;

// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

namespace KS.Misc.Screensaver.Displays
{
	public static class NoiseSettings
	{

		private static int _noiseNewScreenDelay = 5000;
		private static int _noiseDensity = 40;

		/// <summary>
        /// [Noise] How many milliseconds to wait before making the new screen?
        /// </summary>
		public static int NoiseNewScreenDelay
		{
			get
			{
				return _noiseNewScreenDelay;
			}
			set
			{
				if (value <= 0)
					value = 5000;
				_noiseNewScreenDelay = value;
			}
		}
		/// <summary>
        /// [Noise] The noise density in percent
        /// </summary>
		public static int NoiseDensity
		{
			get
			{
				return _noiseDensity;
			}
			set
			{
				if (value < 0)
					value = 40;
				if (value > 100)
					value = 40;
				_noiseDensity = value;
			}
		}

	}

	public class NoiseDisplay : BaseScreensaver, IScreensaver
	{

		private Random RandomDriver;
		private int CurrentWindowWidth;
		private int CurrentWindowHeight;
		private bool ResizeSyncing;

		public override string ScreensaverName { get; set; } = "Noise";

		public override Dictionary<string, object> ScreensaverSettings { get; set; }

		public override void ScreensaverPreparation()
		{
			// Variable preparations
			RandomDriver = new Random();
			CurrentWindowWidth = ConsoleWrapper.WindowWidth;
			CurrentWindowHeight = ConsoleWrapper.WindowHeight;
		}

		public override void ScreensaverLogic()
		{
			double NoiseDense = (NoiseSettings.NoiseDensity > 100 ? 100 : NoiseSettings.NoiseDensity) / 100d;

			Console.BackgroundColor = ConsoleColor.DarkGray;
			ConsoleWrapper.CursorVisible = false;
			ConsoleWrapper.Clear();
			Console.BackgroundColor = ConsoleColor.Black;

			// Select random positions to generate noise
			int AmountOfBlocks = ConsoleWrapper.WindowWidth * ConsoleWrapper.WindowHeight;
			int BlocksToCover = (int)Math.Round(AmountOfBlocks * NoiseDense);
			var CoveredBlocks = new ArrayList();
			while (!(CoveredBlocks.Count == BlocksToCover | ResizeSyncing))
			{
				if (CurrentWindowHeight != ConsoleWrapper.WindowHeight | CurrentWindowWidth != ConsoleWrapper.WindowWidth)
					ResizeSyncing = true;
				if (!ResizeSyncing)
				{
					int CoverX = RandomDriver.Next(ConsoleWrapper.WindowWidth);
					int CoverY = RandomDriver.Next(ConsoleWrapper.WindowHeight);
					ConsoleWrapper.SetCursorPosition(CoverX, CoverY);
					TextWriterColor.WritePlain(" ", false);
					if (!CoveredBlocks.Contains(CoverX.ToString() + ", " + CoverY.ToString()))
						CoveredBlocks.Add(CoverX.ToString() + ", " + CoverY.ToString());
				}
				else
				{
					// We're resizing.
					break;
				}
			}

			// Reset resize sync
			ResizeSyncing = false;
			CurrentWindowWidth = ConsoleWrapper.WindowWidth;
			CurrentWindowHeight = ConsoleWrapper.WindowHeight;
			ThreadManager.SleepNoBlock(NoiseSettings.NoiseNewScreenDelay, ScreensaverDisplayer.ScreensaverDisplayerThread);
		}

	}
}